library/graylog_streams.py

```diff
@@ -381,27 +381,88 @@ def create(module, base_url, headers, index_set_id):
 
     return info['status'], info['msg'], content, url
 
-
-def create_rule(module, base_url, headers):
+def query_rules(module, base_url, headers):
 
     url = "/".join([base_url, module.params['stream_id'], "rules"])
-
+    #raise Exception(url)
     payload = {}
+    field = module.params['field']
+    value = module.params['value']
+    #raise Exception(module.params['field'])
 
-    for key in ['field', 'type', 'value', 'inverted', 'description']:
-        if module.params[key] is not None:
-            payload[key] = module.params[key]
-
-    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='POST', data=module.jsonify(payload))
 
-    if info['status'] != 201:
+    #TODO: GET current rules, compare field and value to determine duplication
+    response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='GET')
+    if info['status'] != 200:
         module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))
 
     try:
         content = to_text(response.read(), errors='surrogate_or_strict')
+        rules = json.loads(content)
+        #raise Exception(rules)
     except AttributeError:
         content = info.pop('body', '')
 
+    if rules is not None:
+        i = 0
+        while i < len(rules['stream_rules']):
+            rule = rules['stream_rules'][i]
+            if field == rule['field'] and value == rule['value']:
+                rule_json = {'rule_id': rule['id']}
+                #raise Exception(rule_json)
+                break
+            else:
+                rule_json = {'rule_id': '0'}
+            i += 1
+    else:
+        raise Exception("No streams returned from Graylog API")
+
+    return info['status'], info['msg'], json.dumps(rule_json), url
+
+    #for key in ['field', 'type', 'value', 'inverted', 'description']:
+    #    if module.params[key] is not None:
+    #        payload[key] = module.params[key]
+
+    #response, info = fetch_url(module=module, url=url, headers=json.loads(headers), method='POST', data=module.jsonify(payload))
+
+    #if info['status'] != 201:
+    #    module.fail_json(msg="Fail: %s" % ("Status: " + str(info['msg']) + ", Message: " + str(info['body'])))
+
+    #try:
+    #    content = to_text(response.read(), errors='surrogate_or_strict')
+    #except AttributeError:
+    #    content = info.pop('body', '')
+
+    #return info['status'], info['msg'], content, url
+
+def create_rule(module, base_url, headers):
+
+    status, message, content, url = query_rules(module, base_url, headers)
+    #raise Exception(f'{status} {message} {content}')
+    query_result = json.loads(content)
+    #raise Exception(result['rule_id'])
+    if query_result['rule_id'] == "0":
+        url = "/".join([base_url, module.params['stream_id'], "rules"])
```

Comment:
Better to use `urllib` methods for constructing URLs.

Reply:
You're right of course but, hell of a can of worms given they way they've set it up originally 😬 . 
I'll fix it though since I need to do the same in the rest of the module.

Reply:
OK, NP. Just a suggestion maybe for next time.

Reply:
I made these changes in most of the functions... the ones I'm actually using. 
I still had to use "/".join for some endpoint paths in order to avoid a lot of repeated URL building code.